# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/workers/ScriptLoader.cpp
# Commit: 68b476066248
# Full Hash: 68b476066248deb9b80b0be44bebb991fdc94a6d
# Author: Yulia Startsev <ystartsev@mozilla.com>
# Date: 2022-12-21 21:21:23
# Regressor Bug: 1247687
# File Overlap Count: 1
# Description:
#   Bug 1247687 - Implement InitModuleLoader methods for WorkerScriptLoader and WorkerGlobalScope; r=jonco
#   
#   This implements a method to initialize the moduleLoader for workers. This will initialize only once, for all worker types (module and classic).
#   
#   Depends on D147324
# ==============================================================================

diff -r c94a9dc60dff -r 68b476066248 dom/workers/ScriptLoader.cpp
--- a/dom/workers/ScriptLoader.cpp	Tue Dec 20 20:56:11 2022 +0000
+++ b/dom/workers/ScriptLoader.cpp	Tue Dec 20 20:56:12 2022 +0000
@@ -411,8 +411,20 @@
   }
 
   nsIGlobalObject* global = GetGlobal();
+  mController = global->GetController();
+  // Set up the module loader, if it has not been yet.
+  // TODO: Implement this for classic scripts when dynamic imports are added.
+  if (aWorkerPrivate->WorkerType() == WorkerType::Module) {
+    InitModuleLoader();
+  }
+}
 
-  mController = global->GetController();
+void WorkerScriptLoader::InitModuleLoader() {
+  mWorkerRef->Private()->AssertIsOnWorkerThread();
+  RefPtr<WorkerModuleLoader> moduleLoader =
+      new WorkerModuleLoader(this, GetGlobal(), mSyncLoopTarget.get());
+  static_cast<WorkerGlobalScopeBase*>(GetGlobal())
+      ->InitModuleLoader(moduleLoader);
 }
 
 void WorkerScriptLoader::CreateScriptRequests(